storage/src/binary/mod.rs

```diff
@@ -0,0 +1,457 @@
+        let tree = Tree::new(&mut hasher, digests.clone()).unwrap();
+
+        // Generate a valid proof for leaf at index 1.
+        let proof = tree.prove(1).unwrap();
+        let mut serialized = proof.serialize();
+
+        // Append an extra byte.
+        serialized.push(0u8);
+        let deserialized = Proof::<Sha256>::deserialize(&serialized);
+        assert!(
+            deserialized.is_none(),
+            "Deserialization should fail with extra data"
+        );
+    }
+
+    #[test]
+    fn test_invalid_proof_modified_hash() {
+        // Build tree
+        let txs = [b"tx1", b"tx2", b"tx3", b"tx4"];
+        let digests: Vec<Digest> = txs.iter().map(|tx| hash(*tx)).collect();
+        let mut hasher = Sha256::default();
+        let tree = Tree::new(&mut hasher, digests.clone()).unwrap();
+        let root = tree.root();
+
+        // Generate a valid proof for leaf at index 2.
+        let mut proof = tree.prove(2).unwrap();
+
+        // Modify the first hash in the proof.
+        proof.hashes[0] = hash(b"modified");
+        let valid = proof.verify(&mut hasher, &digests[2], 2, &root);
+        assert!(
+            !valid,
+            "Verification should fail if a proof hash is tampered with"
+        );
+    }
+
+    #[test]
+    fn test_odd_tree_duplicate_index_proof() {
+        // Build a tree with an odd number of leaves.
+        let txs = [b"tx1", b"tx2", b"tx3"];
+        let digests: Vec<Digest> = txs.iter().map(|tx| hash(*tx)).collect();
+        let mut hasher = Sha256::default();
+        let tree = Tree::new(&mut hasher, digests.clone()).unwrap();
+        let root = tree.root();
+
+        // The tree was built with 3 leaves; index 2 is the last valid index.
+        let proof = tree.prove(2).unwrap();
+
+        // Verification should succeed for the proper index 2.
+        assert!(proof.verify(&mut hasher, &digests[2], 2, &root));
+
+        // Should not be able to generate a proof for an out-of-range index (e.g. 3).
+        assert!(tree.prove(3).is_none());
+
+        // Attempting to verify using an out-of-range index (e.g. 3, which would correspond
+        // to a duplicate leaf that doesn't actually exist) should fail.
+        assert!(
+            !proof.verify(&mut hasher, &digests[2], 3, &root),
+            "Verification should fail for an invalid duplicate leaf index"
+        );
+    }
+
+    #[test]
+    fn test_odd_tree_default_index_proof() {
+        // Build a tree with an odd number of leaves.
+        let txs = [b"tx1", b"tx2", b"tx3"];
+        let digests: Vec<Digest> = txs.iter().map(|tx| hash(*tx)).collect();
+        let mut hasher = Sha256::default();
+        let tree = Tree::new(&mut hasher, digests.clone()).unwrap();
+        let root = tree.root();
+
+        // The tree was built with 3 leaves; index 2 is the last valid index.
+        let mut proof = tree.prove(2).unwrap();
+
+        // Swap sibling with last valid item
+        proof.hashes[0] = digests[2].clone();
```

Comment:
@roberto-bayardo not sure if you have a test that does something like this but might be interesting to consider...

Reply:
Is this to confirm that there's no "leaf-node weakness"?  No don't believe I tested for this explicitly in the MMR.

Reply:
Actually I think there is a leaf-node weakness. usually you will want to use some kind of "domain separation" between nodes & leaves, and I don't see where that might be happening?

Reply:
> the root itself is hashed with the size of the tree

👀

Reply:
Resolved now 🫡 .